### PR TITLE
JS: DB reads as taint sources

### PR DESCRIPTION
**Pull request by kaeluka**
This adds DB reads as taint sources in `AdditionalSources.qll`.

~~**This is not yet complete**, as it's not handling streaming database queries where the result is consumed by a pipe call (eg, knex, spanner, have such features).~~

### CODE REVIEW

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll [hunk 1]
```diff
@@ -57,6 +57,26 @@ module Knex {
       )
     }
 
-    override DataFlow::Node getAQueryArgument() { none() }
+    override DataFlow::Node getAResult() { 
+      (
```
Comment (kaeluka) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:61]: the knex model is a bit unusual in that it classifies both the expression producing the promise, as well as the `await` expression as database accesses. `getAResult`, in this case, follows the implementation of the characteristic predicate. LMK what you think.

File: javascript/ql/lib/semmle/javascript/Concepts.qll [hunk 2]
```diff
@@ -84,6 +84,7 @@ abstract class FileNameSource extends DataFlow::Node { }
 abstract class DatabaseAccess extends DataFlow::Node {
   /** Gets an argument to this database access that is interpreted as a query. */
   abstract DataFlow::Node getAQueryArgument();
+  abstract DataFlow::Node getAResult();
```
Comment (intrigus-lgtm) [inline javascript/ql/lib/semmle/javascript/Concepts.qll:87]: Missing Ql Doc.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 3]
```diff
@@ -604,6 +662,18 @@ private module Spanner {
...
+    }
+
+    override DataFlow::Node getAResult() {
+      PromiseFlow::loadStep(this.getALocalUse(), result, Promises::valueProp())
+      or
+      (this = database().getMember("run").getACall() and result = this.getCallback(_).getParameter(1))
+      or
+      (this = transaction().getMember("run").getACall() and result = this.getCallback(_).getParameter(1))
+      or
+      (this = snapshot().getMember("run").getACall() and result = this.getCallback(_).getParameter(1))
```
Comment (intrigus-lgtm) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:676]: I think you should be able to simplify this using set-literals to
```ql
this = [database(), transaction(), snapshot()].getMember("run").getACall() and result = this.getCallback(_).getParameter(1))
```

File: javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll [hunk 4]
```diff
@@ -52,3 +52,13 @@ class RemoteServerResponse extends HeuristicSource, RemoteFlowSource {
 
   override string getSourceType() { result = "a response from a remote server" }
 }
+
+class DatabaseAccessResultRemoteFlowSource extends HeuristicSource, RemoteFlowSource {
```
Comment (intrigus-lgtm) [inline javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll:56]: Missing Ql Doc.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll (hunk 3 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:676]: clever, thx! :)

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll (hunk 1 above)
Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:61]: It seems reasonable.  
One thing you could do is to split up the class into two (one with the await, and one without). 
That might allow the code to be a bit cleaner.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 5]
```diff
@@ -174,6 +179,26 @@ private module Postgres {
   private class QueryCall extends DatabaseAccess, DataFlow::MethodCallNode {
     QueryCall() { this = [client(), pool()].getMember("query").getACall() }
 
+    override DataFlow::Node getAResult() {
+      (
+        this.(DataFlow::CallNode).getNumArgument() = 2 and
+        result = this.(DataFlow::CallNode)
+        .getCallback(1)
+        .getParameter(1)
+      )
```
Comment (kaeluka) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:188]: remove those parens (and in other places)

**Comment by kaeluka** [comment]
Had a chat with @erik-krogh, I'll for now not model streaming APIs. We might choose to do that later.

**Comment by kaeluka** [comment]
left to be done:
 - implement `getResults` for rest of DBs.
 - query new results on lgtm to evaluate how noisy those sources will be.
 - bring test up to current idiom (using comments on lines that should be found).
 - performance evaluation.

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll (hunk 1 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/Knex.qll]: I did split it up!

**Comment by kaeluka** [comment]
> * query new results on lgtm to evaluate how noisy those sources will be.

I've now used lgtm to [select all database accesses](https://lgtm.com/query/2212660374859748682/).
Octobox and strider are the projects with the highest number of database accesses. I've downloaded those locally and am running my changes against them. The result is that there's on average < 1 result per database access (because some database accesses, like data deletions or updates, do not use a result at all).

This means that there's no surprises. Does that make sense to you, @erik-krogh?

**Comment by erik-krogh** [comment]
> This means that there's no surprises. Does that make sense to you, @erik-krogh?

That makes sense to me.  
I also took at look at [your evaluation where you added these DB reads as taint sources](https://github.com/github/codeql-dca-main/tree/data/kaeluka/pr-7474-73a85c5__nightly__security-extend/reports).  
Performance seems to be within the margin of error.  

But the 4 new results all look like FPs to me.  
So I think it's the right decision to have the new source as a heuristic source.

File: javascript/ql/lib/semmle/javascript/Concepts.qll [hunk 6]
```diff
@@ -84,6 +84,17 @@ abstract class FileNameSource extends DataFlow::Node { }
 abstract class DatabaseAccess extends DataFlow::Node {
   /** Gets an argument to this database access that is interpreted as a query. */
   abstract DataFlow::Node getAQueryArgument();
+
+  /** Gets a node to which a result of the access may flow. */
+  abstract DataFlow::Node getAResult();
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/Concepts.qll:89]:
```suggestion
  DataFlow::Node getAResult() {
    none() // Overridden in subclass
  }
```

We try to keep backwards comparability with QL code written by others.  
Adding a new abstract predicate will result in a compilation error for any existing implementation. 

An alternative pattern we often use is to add the predicate with a default `none()` implementation.

File: javascript/ql/lib/semmle/javascript/Concepts.qll [hunk 7]
```diff
@@ -84,6 +84,17 @@ abstract class FileNameSource extends DataFlow::Node { }
...
+
+  /** Might the data returned as a JSON object like document databases do?
+   * */
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/Concepts.qll:92]:
```suggestion
  /** 
   * Holds if the data returned is can be a user-controlled object, such as a JSON object parsed from user-controlled data. 
   * */
```
I rewrote the QLDoc. Predicates without a result should generally start with "Holds ...". 
I took some inspiration from `RemoteFlowSource::isUserControlledObject()`.

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll [hunk 8]
```diff
@@ -46,17 +46,38 @@ module Knex {
...
   }
 
-  /** A call that triggers a SQL query submission. */
-  private class KnexDatabaseAccess extends DatabaseAccess {
-    KnexDatabaseAccess() {
+  /** A call that triggers a SQL query submission by calling then/stream/asCallback. */
+  private class KnexDatabaseCallback extends DatabaseAccess, DataFlow::CallNode {
+    KnexDatabaseCallback() {
       this = knexObject().getMember(["then", "stream", "asCallback"]).getACall()
+    }
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:53]:
```suggestion
    string member;
    KnexDatabaseCallback() {
      member = ["then", "stream", "asCallback"] and
      this = knexObject().getMember(member).getACall()
    }
```
You can simplify `getAResult()` if you move `member` to a field.

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll [hunk 9]
```diff
@@ -46,17 +46,38 @@ module Knex {
...
+
+    override DataFlow::Node getAResult() { 
+      this = knexObject().getMember("then").getACall() and
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:56]:
```suggestion
      member = "then" and
```
(Belongs with the above suggestion)

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll [hunk 10]
```diff
@@ -46,17 +46,38 @@ module Knex {
...
+        result = this.getCallback(0).getParameter(0)
       or
-      exists(AwaitExpr await |
-        this = await.flow() and
-        await.getOperand() = knexObject().getAUse().asExpr()
-      )
+        this = knexObject().getMember("asCallback").getACall() and
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:59]:
```suggestion
      member = "asCallback" and
```
(Belongs with the above suggestion)

It also look like you need to run the autoformatter on this code (Ctrl + Shift + P -> `Format document`).

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll [hunk 11]
```diff
@@ -46,17 +46,38 @@ module Knex {
...
+    override DataFlow::Node getAQueryArgument() {
+      none()
+    }
+  }
+
+  private class KnexDatabaseAwait extends DatabaseAccess, DataFlow::ValueNode {
+    AwaitExpr enclosingAwait;
+    KnexDatabaseAwait() {
+      this = enclosingAwait.flow() and
+      enclosingAwait.getOperand() = knexObject().getAUse().asExpr()
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:72]:
```suggestion
    KnexDatabaseAwait() {
      exists(AwaitExpr enclosingAwait | this = enclosingAwait.flow() | 
        enclosingAwait.getOperand() = knexObject().getAUse().asExpr()
      )
```

You have a field that is not used outside of the characteristic predicate. The field can be inlined into an `exists(..)`.

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll [hunk 12]
```diff
@@ -537,14 +542,22 @@ private module Mongoose {
...
       f.getQueryArgument().getARhs() = result
     }
+
+    override DataFlow::Node getAResult() {
+      result = this.getCallback(this.getNumArgument() - 1).getParameter(1)
+    }
   }
 
-  class ExplicitQueryEvaluation extends DatabaseAccess {
+class ExplicitQueryEvaluation extends DatabaseAccess {
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:551]:
```suggestion
  class ExplicitQueryEvaluation extends DatabaseAccess {
```
Looks like a spurious whitespace change.  
I also think the autoformatter will fail on this.

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll [hunk 13]
```diff
@@ -537,14 +542,22 @@ private module Mongoose {
...
     ExplicitQueryEvaluation() {
       // explicit execution using a Query method call
       Query::getAMongooseQuery().getMember(["exec", "then", "catch"]).getACall() = this
     }
 
+    override DataFlow::Node getAResult() {
+      result = this.(DataFlow::CallNode).getCallback(_).getParameter(1)
+    }
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:559]: 2 things:  

1) You can move `DataFlow::CallNode` to the `extends` clause, that way you don't have to do an line cast here. 
2) It looks suspicious to me that you are selecting the second parameter of every callback.  
    You are saying that `foo` in `query.then((foo) => ...)` is **not** a result?

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll [hunk 14]
```diff
@@ -623,6 +640,10 @@ private module MarsDB {
 
     string getMethod() { result = method }
 
+    override DataFlow::Node getAResult() {
+      PromiseFlow::loadStep(this.(DataFlow::SourceNode).getALocalUse(), result, Promises::valueProp())
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:644]: If you add `DataFlow::CallNode` to the `extends` clause, then you don't need an inline cast here.

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll [hunk 15]
```diff
@@ -638,6 +659,10 @@ private module MarsDB {
       )
     }
 
+    override DataFlow::Node getAResult() {
+      PromiseFlow::loadStep(this.(DataFlow::SourceNode).getALocalUse(), result, Promises::valueProp())
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:663]: If you add `DataFlow::CallNode` to the `MarsDBAccess` class, then this inline cast becomes redundant. 

There is a bunch of others that are similar, I haven't added a comment on the rest.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 16]
```diff
@@ -174,6 +179,16 @@ private module Postgres {
...
+    override DataFlow::Node getAResult() {
+      this.(DataFlow::CallNode).getNumArgument() = 2 and
+      result = this.(DataFlow::CallNode).getCallback(1).getParameter(1)
+      or
+        this.(DataFlow::CallNode).getNumArgument() = 1 and
+        result = this.getAMethodCall("then").getCallback(0).getParameter(0)
+      or
+        PromiseFlow::loadStep(this.getALocalUse(), result, Promises::valueProp())
+    }
+
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:191]: I think the autoformatter will fail on this.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 17]
```diff
@@ -580,6 +618,16 @@ private module Spanner {
     result = API::Node::ofType("@google-cloud/spanner", "Transaction")
   }
 
+  API::Node snapshot() {
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:621]: QLDoc?

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 18]
```diff
@@ -637,6 +697,13 @@ private module Spanner {
       this = v1SpannerClient().getMember(["executeSql", "executeStreamingSql"]).getACall()
     }
 
+    override DataFlow::Node getAResult() {
+      (
+        this = v1SpannerClient().getMember("executeSql").getACall() and
+        result = this.(DataFlow::CallNode).getCallback(_).getParameter(1)
+      )
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:704]:
```suggestion
      this = v1SpannerClient().getMember("executeSql").getACall() and
      result = this.(DataFlow::CallNode).getCallback(_).getParameter(1)
```
Superfluous parentheses.

**Review by erik-krogh — COMMENTED** [review]
It's generally looking good, I just got a bunch of minor comments.  

The autoformatter is failing for a lot of your files.  

I have the following installed as a `pre-commit` hook. That hook prevents me from commiting QL code that is not formatted correctly: 

```bash
#!/bin/bash
exec 1>&2
exitVal=0
for f in $(git diff --cached --relative --name-only)
do
  filename="${f##*/}"
  extension="${filename##*.}"
  p="$PWD/$f";
  if [[ -f $p ]] && ([ $extension == "ql" ] || [ $extension == "qll" ])
  then
    codeql query format --check-only "$p"
    retVal=$?
    if [[ $retVal -ne 0 ]]
    then
      exitVal=1
    fi
  fi
done
exit $exitVal
```

File: javascript/ql/lib/semmle/javascript/Concepts.qll (hunk 6 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/Concepts.qll:89]: perfect, will do

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll (hunk 17 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:620]: 👍

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll (hunk 18 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:700]: 👍

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll (hunk 15 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll]: thx!

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll (hunk 16 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll]: I had no idea the autoformatter existed 🙈

File: javascript/ql/lib/semmle/javascript/Concepts.qll (hunk 7 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/Concepts.qll]: 👍

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll (hunk 11 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/Knex.qll]: 👍

File: javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll (hunk 13 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/NoSQL.qll:559]: 1. is locally resolved, but 2. fixed this (good catch!) and added a unit test case for it, too.

**Comment by kaeluka** [comment]
I have now fixed those issues. Thanks for the review, @erik-krogh — sorry that you had to deal with so much whitespace stuff. I now have a commit hook like your's that'll prevent this in the future :)

**Comment by erik-krogh** [comment]
> I have now fixed those issues. Thanks for the review, @erik-krogh — sorry that you had to deal with so much whitespace stuff. I now have a commit hook like your's that'll prevent this in the future :) 

Great :+1:  
Now you got a merge conflict (I think that's from Asgers MaD work).  
I'll take another look when that has been resolved.

**Comment by kaeluka** [comment]
Alright, @erik-krogh, I've rebased, resolved the conflicts, pushed.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 19]
```diff
@@ -540,3 +598,145 @@ private module SpannerCsv {
     }
   }
 }
+
+/**
+ * Provides classes modelling the Google Cloud Spanner library.
```
Comment (esbena) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:603]:
```suggestion
 * Provides classes modeling the Google Cloud Spanner library.
```
(as suggested by code scanning)

Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:603]: *sigh* I'll need a spell checker for VS Code. Thx!

File: javascript/ql/lib/semmle/javascript/frameworks/Knex.qll [hunk 20]
```diff
@@ -46,17 +46,35 @@ module Knex {
...
+      result = this.getCallback(0).getParameter(1)
+    }
+
+    override DataFlow::Node getAQueryArgument() { none() }
+  }
+
+  private class KnexDatabaseAwait extends DatabaseAccess, DataFlow::ValueNode {
+    KnexDatabaseAwait() {
+      exists(AwaitExpr enclosingAwait | this = enclosingAwait.flow() |
+        enclosingAwait.getOperand() = knexObject().getAUse().asExpr()
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/Knex.qll:72]: I think you're right. I tried adding `PromiseFlow::flowStep`, but it does not immediately work. I'll take a closer look.

File: javascript/ql/lib/semmle/javascript/Concepts.qll [hunk 21]
```diff
@@ -84,6 +84,21 @@ abstract class FileNameSource extends DataFlow::Node { }
...
   /** Gets an argument to this database access that is interpreted as a query. */
   abstract DataFlow::Node getAQueryArgument();
+
+  /** Gets a node to which a result of the access may flow. */
+  DataFlow::Node getAResult() {
+    none() // Overridden in subclass
+  }
+
+  /**
+   * Holds if the data returned is can be a user-controlled object,
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/Concepts.qll:94]: 👍

**Review by esbena — COMMENTED** [review]
LGTM. But I'll defer to Erik for final approval since mine was more of a drive-by review.

**Comment by kaeluka** [comment]
Thanks, Esben! Makes sense to wait on @erik-krogh!

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 22]
```diff
@@ -540,3 +598,145 @@ private module SpannerCsv {
     }
   }
 }
+
+/**
+ * Provides classes modeling the Google Cloud Spanner library.
+ */
+private module Spanner {
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:605]: This entire module is mostly copy-paste from the model that was "deprecated" by models-as-data.  

You could try to use models-as-data, by extending the rows in `SpannerSinks` with new rows, where you invent some new sink-type (like `database-access`).  
In `AdditionalSources.qll` you can then access the new sinks like: `ModelOutput::getASinkNode("database-access").getARhs()`.   

Models-as-data has an `Awaited` path token that you can use. 

The rest looks fine. 

Another comment: I usually don't squash commits after un-drafting a PR. Having multiple commits makes it easier to review.

Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:605]: > You could try to use models-as-data, by extending the rows in SpannerSinks with new rows, where you invent some new sink-type (like database-access).

You mean create a new `SpannerSources` module, right? I've been looking at the MaD work (the details of which I didn't yet know all that well), and it looks promising. I'll start working on that now. Thank you.

> Another comment: I usually don't squash commits after un-drafting a PR. Having multiple commits makes it easier to review.

Thanks for pointing that out, I'll happily match what others are doing :) Is it customary to do a final squash before merging?

Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:605]: > You mean create a new `SpannerSources` module, right? I've been looking at the MaD work (the details of which I didn't yet know all that well), and it looks promising. I'll start working on that now. Thank you.

Yes, right. 
Probably just a new class `class SpannerSources extends ModelInput::SourceModelCsv`. 
And then use it like: `ModelOutput::getASourceNode("database-access").getAnImmediateUse()`. 

> Thanks for pointing that out, I'll happily match what others are doing :) Is it customary to do a final squash before merging? 

No it's not. 
The GitHub web UI shows a link from a commit to the PR that included the PR.  
That feature makes squash commits somewhat redundant.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 23]
```diff
@@ -574,7 +574,10 @@ private module SpannerCsv {
           "@google-cloud/spanner;;@google-cloud/spanner;;Member[Spanner]",
           "@google-cloud/spanner;Database;@google-cloud/spanner;;ReturnValue.Member[instance].ReturnValue.Member[database].ReturnValue",
           "@google-cloud/spanner;v1.SpannerClient;@google-cloud/spanner;;Member[v1].Member[SpannerClient].Instance",
-          "@google-cloud/spanner;Transaction;@google-cloud/spanner;Database;Member[runTransaction,runTransactionAsync].Argument[0..1].Parameter[1]",
+          "@google-cloud/spanner;Transaction;@google-cloud/spanner;Database;Member[runTransaction,runTransactionAsync,getTransaction].Argument[0..1].Parameter[1]",
+          "@google-cloud/spanner;Transaction;@google-cloud/spanner;Database;Member[getTransaction].ReturnValue.Awaited",
```
Comment (kaeluka) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:578]: This line had been overlooked before. Added it.

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll (hunk 22 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll]: I just pushed commit 09a28c428c25714b1b10cd4312c5e989152833e6. Looks good to you?

> No it's not.
The GitHub web UI shows a link from a commit to the PR that included the PR.
That feature makes squash commits somewhat redundant.

Alright, will do :) Thanks for pointing that out.

File: javascript/ql/test/library-tests/Security/heuristics/sources.js [hunk 24]
```diff
@@ -152,6 +153,7 @@ function sink(data) {}
         txn.run('SELECT * FROM users', function (err, users) {
             sink(users); // NOT OK
         });
+        txn.commit(function () {});
```
Comment (erik-krogh) [inline javascript/ql/test/library-tests/Security/heuristics/sources.js:156]: Why has this line been added to the test? 
Was it to make the code in the test more realistic? (a perfectly valid reason).

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 25]
```diff
@@ -539,4 +600,22 @@ private module SpannerCsv {
...
+        [
+          "Member[executeSql].Argument[0..].Parameter[1]",
+          "Member[executeSql].ReturnValue.Awaited.Member[0]", "Member[run].ReturnValue.Awaited",
+          "Member[run].Argument[0..].Parameter[1]",
+        ]
+    }
+
+    override predicate row(string row) {
+      row =
+        "@google-cloud/spanner;" + spannerClass() + ";" + resultPath() + ";database-access-result"
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:618]:
```suggestion
        "@google-cloud/spanner;" + this.spannerClass() + ";" + this.resultPath() +
          ";database-access-result"
```
Fix code-scanning recommendation, (and an attempt to keep the auto-formatter happy).

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 26]
```diff
@@ -420,6 +442,11 @@ private module MsSql {
       mssql().getMember("query").getAUse() = DataFlow::valueNode(astNode.getTag())
     }
 
+    override DataFlow::Node getAResult() {
+      PromiseFlow::loadStep(this.(DataFlow::SourceNode).getALocalUse(), result,
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:446]: A `DataFlow::valueNode` where the `astNode` is of type `TaggedTemplateExpr` is always a `DataFlow::SourceNode` (see the `DefaultRange` class in `Sources.qll`).  
So you should be able to add `DataFlow::SourceNode` on top of the other classes in the extends clause. 

(Just to get rid of the inline cast to `DataFlow::SourceNode`).

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll [hunk 27]
```diff
@@ -491,6 +524,31 @@ private module Sequelize {
     }
   }
 
+  /** Gets an import of the `sequelize` module or one that re-exports it. */
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:527]: Now that you've converted the `Spanner` model to use MaD, could you do the same with your `sequelize` model?

**Review by erik-krogh — COMMENTED** [review]
We are in the final stretch, almost there :tada: 

But I still have some comments.

File: javascript/ql/test/library-tests/Security/heuristics/sources.js (hunk 24 above)
Comment (kaeluka) [reply javascript/ql/test/library-tests/Security/heuristics/sources.js:156]: yes, realistic!

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll (hunk 25 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll:618]: have done so :)

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll (hunk 26 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll]: done!

File: javascript/ql/lib/semmle/javascript/frameworks/SQL.qll (hunk 27 above)
Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/frameworks/SQL.qll]: the second time around, it was much easier :)

File: javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll [hunk 28]
```diff
@@ -52,3 +52,30 @@ class RemoteServerResponse extends HeuristicSource, RemoteFlowSource {
...
+
+  override predicate isUserControlledObject() { any() }
+}
+
+/**
+ * A remote flow source originating from a database access.
+ */
+private class RemoteFlowSourceFromDBAccess extends RemoteFlowSource, HeuristicSource {
+  RemoteFlowSourceFromDBAccess() {
+    this = ModelOutput::getASourceNode("database-access-result").getAUse()
```
Comment (erik-krogh) [inline javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll:72]:
```suggestion
    this = ModelOutput::getASourceNode("database-access-result").getAUse()
    or
    exists(DatabaseAccess dba | this = dba.getAResult())
```

One last comment, couldn't you combine the two new classes in `AdditionalSources.qll`?

Comment (kaeluka) [reply javascript/ql/lib/semmle/javascript/heuristics/AdditionalSources.qll]: sure, done

**Review by erik-krogh — APPROVED** [review]
:tada:

**Comment by kaeluka** [comment]
Thanks for the super helpful reviews, @erik-krogh!

**Pull request merged**